abci/app.go

```diff
@@ -0,0 +1,378 @@
+
+		return &resp, nil
+	}
+	// Retrieve all message sent by the sender
+	messages, err := model.GetMessagesBySender(app.state.DB, sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the messages to JSON and return as query result
+	resultBytes, err := json.Marshal(messages)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.Log = string(resultBytes)
+	resp.Value = resultBytes
+
+	return &resp, nil
+}
+
+func (app ForumApp) CheckTx(ctx context.Context, checktx *abci.RequestCheckTx) (*abci.ResponseCheckTx, error) {
+	// Parse the tx message
+	msg, err := model.ParseMessage(checktx.Tx)
+	if err != nil {
+		fmt.Printf("failed to parse transaction message checktx: %v\n", err)
+		return &abci.ResponseCheckTx{Code: CodeTypeInvalidTxFormat, Log: "Invalid transaction format"}, err
+	}
+	fmt.Println("Searching for sender ... ", msg.Sender)
+	u, err := app.state.DB.FindUserByName(msg.Sender)
+
+	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+		fmt.Println("problem in check tx: ", string(checktx.Tx))
+		return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
+	}
+
+	if err == badger.ErrKeyNotFound {
+		fmt.Println("Not found user :", msg.Sender)
+	}
+	if u != nil && u.Banned {
+		return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
+	}
+	fmt.Println("Check tx success for ", msg.Message, " and ", msg.Sender)
+	return &types.ResponseCheckTx{Code: CodeTypeOK}, nil
+}
+
+// Consensus Connection
+// Initialize blockchain w validators/other info from CometBFT
+func (app ForumApp) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	for _, v := range req.Validators {
+		app.updateValidator(v)
+	}
+	appHash := make([]byte, 8)
+	binary.PutVarint(appHash, app.state.Size)
+	req.ConsensusParams.Abci.VoteExtensionsEnableHeight = 1
+	return &abci.ResponseInitChain{ConsensusParams: req.ConsensusParams, AppHash: appHash}, nil
+}
+
+func (app *ForumApp) PrepareProposal(_ context.Context, proposal *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	fmt.Println("entered prepareProp")
+
+	voteExtensionCurseWords := app.getWordsFromVe(proposal.LocalLastCommit.Votes)
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			if err != nil {
+				proposedTxs = append(proposedTxs, tx)
+			}
+			// Adding the curse words from vote extensions too
+			if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+				proposedTxs = append(proposedTxs, tx)
+			} else {
+				banTx := model.BanTx{UserName: msg.Sender}
+				bannedUsersString[msg.Message] = struct{}{}
+				resultBytes, err := json.Marshal(banTx)
+				if err == nil {
+					finalProposal = append(finalProposal, resultBytes)
+				}
```

Comment:
So what do we do if this error is not `nil`?

Reply:
We should probably panic because if we can't create this TX then we will not ban who we should ban. ToDo